distribution/std-lib/Base/src/System/File.enso

```diff
@@ -0,0 +1,261 @@
+import Base.System.File.Option
+
+export Base.System.File.Option
+
+polyglot java import java.nio.file.NoSuchFileException
+polyglot java import java.nio.file.AccessDeniedException
+polyglot java import java.io.IOException
+
+type Error
+    type No_Such_File_Error file
+    type Access_Denied_Error file
+    type Io_Error message
+
+## PRIVATE
+
+   Utility method for rewrapping Java exceptions into Enso panics.
+rethrow_java file exception =
+    case exception of
+        Polyglot_Error exc ->
+            if Java.is_instance exc NoSuchFileException then
+                Panic.throw (No_Such_File_Error file)
+            if Java.is_instance exc AccessDeniedException then
+                Panic.throw (Access_Denied_Error file)
+            if Java.is_instance exc IOException then
+                Panic.throw (Io_Error (exc.getMessage []))
+            Panic.throw exception
+        _ -> Panic.throw exception
+
+## PRIVATE
+
+   Utility method for running an action with Java exceptions mapping.
+with_java_exception_handling file ~action =
+    err = Panic.recover action
+    r = err.catch (here.rethrow_java file)
+    r
+
+## PRIVATE
+
+   Utility method for closing primitive Java streams. Provided to avoid
+   accidental scope capture with `Managed_Resource` finalizers.
+close_stream : Any -> Unit
+close_stream stream =
+    stream.close []
+    Unit
+
+## An output stream, allowing for interactive writing of contents into an
+   open file.
+type Output_Stream
+    type Output_Stream file stream_resource
+
+    ## Writes a vector of bytes into the file at the current stream position.
+    write_bytes : Vector -> Unit
+    write_bytes contents = Managed_Resource.with this.stream_resource java_stream->
+        here.with_java_exception_handling this.file <|
+            java_stream.write [contents.to_array]
+            java_stream.flush []
+        Unit
+
+    ## Closes this stream.
+
+       Even though Streams are closed automatically upon garbage collection, it
+       is still advised to close streams manually if they are not used within
+       a bracket pattern.
+    close : Unit
+    close = Managed_Resource.finalize this.stream_resource
+
+## An input stream, allowing for interactive reading of contents from an open
+   file.
+type Input_Stream
+    type Input_Stream file stream_resource
+
+    ## Reads all the bytes in this file into a vector of bytes.
+    read_all_bytes : Vector
+    read_all_bytes = Managed_Resource.with this.stream_resource java_stream->
+        here.with_java_exception_handling this.file <|
+            Vector.from_polyglot_array (java_stream.readAllBytes [])
+
+    ## Reads _up to_ the provided number of bytes from the stream.
+
+       Makes a best-effort to read as many bytes as provided, however fewer
+       bytes may be read, if end of stream is encountered.
+
+       The length of the returned vector is the same as the number of bytes
+       read.
+    read_n_bytes : Integer -> Vector
```

Comment:
> Why not just read_bytes?

Because we have `read_bytes` on `File` that is the equivalent of this object's `read_all_bytes` so I wanted to avoid confusion.

> The n arg should also default to 1, I think.

That seems like a useless default IMO. If the arg is `1` you don't really want a Vector, so if you find it useful, we could have a `read_byte` that just returns the byte.